### DISCUSSION TITLE
4.0 CharacterBody's Moving Platform layers: Should there be ceilings layers?

### DISCUSSION THREAD

**Discussion by MewPurPur** [godotengine/godot-proposals#4572] [General Discussions]
I was making a 2D Platformer and I found a need for ceiling layers for moving platforms, like how there are layers for floors and walls:

![image](https://user-images.githubusercontent.com/85438892/169658098-7fa89ad0-adbc-401d-b706-f5ca3711e93e.png)

My idea was that it would allow for characters to 
* Stick to the ceiling of a moving platform & potentially apply its velocity on leave.
* Get pushed if they bonk their head on a ceiling.

Like so:

https://user-images.githubusercontent.com/85438892/169657942-8af297d2-3f9e-4552-ba4d-c574524d858c.mp4

I can see how this could seem niche and useless for most people. While I do have a valid use case, it's rare and I'll probably end up using a custom implementation for moving platforms anyway, because my movement mechanics are complex.

So I opened this discussion to get some opinions about whether this should be implemented & is worth the time. Please react with thumbs up or down and please don't be shy to do, even if your opinion winds up being the minority.

Also, if there are any questions about the technical implementation, please voice that. There were some of those in [the proposal I closed in favor of this discussion](https://github.com/godotengine/godot-proposals/issues/4412).

**Comment by Mickeon** [comment]
It is admittedly fairly situational, but in and of itself, seeing "Floor" and "Wall"  but not "Ceiling" Layers feels like something is missing.

It could be useful for heavily physics-based platformers or mechanics, but I do struggle to think of any concrete example, so it really could be **way** too situational.
